src/main/go/org/jenkinsci/plugins/durabletask/heartbeat-launcher.go

```diff
@@ -0,0 +1,235 @@
+package main
+
+import (
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"os/exec"
+	"os/signal"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"golang.org/x/sys/unix"
+)
+
+var mainLogger *log.Logger
+var hbLogger *log.Logger
+var launchLogger *log.Logger
+var scriptLogger *log.Logger
+
+func checkIfErr(process string, err error) bool {
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "%v: %v\n", strings.ToUpper(process), err.Error())
+		return true
+	}
+	return false
+}
+
+func checkScriptErr(err error) bool {
+	return loggerIfErr(scriptLogger, err)
+}
+
+func loggerIfErr(logger *log.Logger, err error) bool {
+	if err != nil {
+		logger.Println(err.Error())
+		return true
+	}
+	return false
+}
+
+// Catch a termination signal to allow for a graceful exit (i.e. no zombies)
+func signalCatcher(sigChan chan os.Signal) {
+	for sig := range sigChan {
+		mainLogger.Printf("(sig catcher) caught: %v\n", sig)
+	}
+}
+
+func launcher(wg *sync.WaitGroup, exitChan chan bool, cookieName string, cookieVal string,
+	interpreter string, scriptPath string, resultPath string, outputPath string) {
+
+	defer wg.Done()
+	defer signalFinished(exitChan)
+
+	var scriptCmd *exec.Cmd
+	if interpreter != "" {
+		scriptCmd = exec.Command(interpreter, "-xe", scriptPath)
+	} else {
+		scriptCmd = exec.Command(scriptPath)
+	}
+	cookieString := fmt.Sprintf("%v=%v", cookieName, cookieVal)
+	scriptCmd.Env = append(os.Environ(),
+		cookieString)
+
+	if outputPath != "" {
+		// capturing output
+		outputFile, err := os.Create(outputPath)
+		if checkScriptErr(err) {
+			exitLauncher(-2, resultPath)
+			return
+		}
+		defer outputFile.Close()
+		scriptCmd.Stdout = outputFile
+	} else {
+		scriptCmd.Stdout = scriptLogger.Writer()
+	}
+	if outputPath != "" {
+		// capturing output
+		scriptCmd.Stderr = scriptLogger.Writer()
+	} else {
+		// Note: pointing to os.Stdout will not capture all err logs
+		scriptCmd.Stderr = scriptCmd.Stdout
+	}
+	// Create new session
+	scriptCmd.SysProcAttr = &unix.SysProcAttr{Setsid: true}
+	for i := 0; i < len(scriptCmd.Args); i++ {
+		launchLogger.Printf("args %v: %v\n", i, scriptCmd.Args[i])
+	}
+	err := scriptCmd.Start()
+	if checkScriptErr(err) {
+		exitLauncher(-2, resultPath)
+		return
+	}
+	pid := scriptCmd.Process.Pid
+	launchLogger.Printf("launched %v\n", pid)
+	err = scriptCmd.Wait()
```

Comment:
It is good that you are calling `Setsid` but I think you also need to fork here somewhere. Otherwise the Golang wrapper as launched by the JVM will not exit until the user process exits, whereas we would like the JVM process handle (reflected in Jenkins APIs as a `Proc`) to exit as soon as the user process has been launched, typically ~100ms after the `sh` step starts. See #98 in which I am enforcing this behavior in tests.

Reply:
It was my understanding that the idea forking, at least in the traditional sense, is not really done in golang. The closest thing you can get to calling a traditional fork in golang would be an `execve`.  Instead, I'm using the `exec` package to spin up a new golang thread.

In this case, I want the golang wrapper to _not_ exit because it is in charge of performing the heartbeat while the  launched script is still running.  I think I might be missing something here?

Reply:
> forking, at least in the traditional sense, is not really done in golang

True (they want you to use external tooling to daemonize) but it seems [it is possible anyway](https://github.com/sevlyar/go-daemon).

> I really want the golang wrapper to _not_ exit because it is in charge of performing the heartbeat

_Some_ Go process must remain running in order to update the heartbeat, redirect stdio, and capture the user process’ exit code. Just not the one that the JVM originally launched.

Reply:
I've been trying to wrestle with a solution in the code as well as in my head. There seem to be two goals here at conflict with one another, and I wished I realized this sooner so I could ask this question in PR-98. 

One goal is being able to run with no zombies in a docker container that does not run `tini`, ie, no zombie reaping for processes with `PPID` 1.  Another goal is the requirement of launching the wrapper process in the background (PR-98).  Any `Proc` that then launches a child into the background, and then exits, will cause that child process to have a `PPID` of 1. And if we are running in a docker container with no `init` process, we have now created a zombie.

I can verify with the parameterized tests in this PR that the changes made in PR-98 will create a zombie in the `simple` docker container. Specifically, the `justSlow`, `reboot`, and `watch` unit test. In other words, anytime a shell script runs long enough, a zombie will be created.

If this assumption is correct, it would seem that the problem that PR-98 is solving is more likely to occur than running out of PIDs? 🤞 And that we just recommend users to start their containers with `--tini`?

Reply:
Bear with me here, as I am not deeply familiar with this area, nor with all the changes you have been making in this PR.

> I can verify with the parameterized tests in this PR that the changes made in PR-98 will create a zombie in the `simple` docker container.

Do you mean to say that #98 introduced a regression? And if so, why would `BourneShellScriptTest.runWithCommandLauncher` in `master` not have caught it?

> we just recommend users to start their containers with `--tini`?

I do not think we can do that without at least some sort of diagnostic warning. Specifically in Kubernetes [there is no such option](https://stackoverflow.com/a/51172104/12916), though it is unclear to me what the practical impact would be in various implementations. Jenkins `sh` steps are sufficiently weird that standard guidelines do not apply.

By the way, when testing in conjunction with the `kubernetes` plugin, it is critical to test both inside and outside the `container` step, as these do entirely different things: outside, the agent JVM is starting the user script using `java.lang.Process` just like a traditional “static” agent; inside, the agent JVM is only used to tail the log file and the like, and the actual process launch is done via the equivalent of `kubectl exec`.

(Quite by coincidence, I just now stumbled across [this](https://ahmet.im/blog/minimal-init-process-for-containers/).)

Reply:
So I do believe #98 introduced a regression, but the reason that `runWithCommandLauncher` did not catch it was because the script was not running long enough.  If you extend the sleep to `60` seconds, for example, the test should fail.  The only reason I caught it was because in my parameterized tests, I run the zombie check on all the tests, so it caught the zombie on the longer running tests. 
Yeah, as long as we have some sort of way to let developers know.  We need some sort of PPID 1 process reaping if we are going to be launching anything in the background without waiting for it to complete.

Reply:
@jglick would it be reasonable then to remove zombie checks from the simple command launcher (since we know that backgrounding a process and then exiting will leave one). And then just adding the diagnostic warning that you mentioned saying that some `init` process will be required if you are running in containers?

Reply:
I guess. I am still trying to figure out what the implications would be, especially for Kubernetes users.

Reply:
Tests have been moved to #103 
Zombie tests are now ignored in the `SIMPLE`/`NO_INIT` platform.
See [JENKINS-58656](https://issues.jenkins-ci.org/browse/JENKINS-58656) for tracking on this issue